### ISSUE TITLE
[naming-convention] prefix wrongly trimmed in snake_case

### ISSUE THREAD

**Issue by haoadoresorange** [typescript-eslint/typescript-eslint#3476]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/naming-convention": [{
                selector: "variable",
                types: ["boolean"],
                format: ["snake_case"],
                prefix: ["is", "should", "has", "can", "did", "will"],
            }]
  }
}
```

```TS
let is_sub_state_changed = false;
```

**Expected Result**

`is_state_changed` correct

**Actual Result**

`isstated_changed` correct

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.22.0` |
| `@typescript-eslint/parser`        | `4.22.0` |
| `TypeScript`                       | `4.2.4` |
| `ESLint`                           | `7.24.0` |
| `node`                             | `14` |

**Comment by bradzacher** [comment]
The prefix logic is simply built to check the prefix.

So the prefix isn't `is`, the prefix is `is_`

This is why if you want to prefix a camel case name you actually have to use Pascal case - because `is` is trimmed, leaving an upper-case first name.

This is on purpose. For example if you wanted to enforce your variable names started with the letter `a` but without the underscore, you can only do that with the current logic.

We don't special case formats because each one requires its own specific logic. And custom formats are completely unhandleable.
